services/autoscaling/src/simcore_service_autoscaling/utils_aws.py

```diff
@@ -2,105 +2,202 @@
+    client = None
+    try:
+        client = boto3.client(
+            "ec2",
+            endpoint_url=settings.EC2_ENDPOINT,
+            aws_access_key_id=settings.EC2_ACCESS_KEY_ID,
+            aws_secret_access_key=settings.EC2_SECRET_ACCESS_KEY,
+            region_name=settings.EC2_REGION_NAME,
+        )
+        yield client
+    finally:
+        if client:
+            client.close()
+
+
+def get_ec2_instance_capabilities(
+    settings: EC2Settings, instance_settings: EC2InstancesSettings
+) -> list[EC2Instance]:
+    with ec2_client(settings) as ec2:
+        instance_types = ec2.describe_instance_types(
+            InstanceTypes=instance_settings.EC2_INSTANCES_ALLOWED_TYPES
+        )
+
+    list_instances: list[EC2Instance] = []
+    for instance in instance_types.get("InstanceTypes", []):
+        with contextlib.suppress(KeyError):
+            list_instances.append(
+                EC2Instance(
+                    name=instance["InstanceType"],
+                    cpus=instance["VCpuInfo"]["DefaultVCpus"],
+                    ram=parse_obj_as(
+                        ByteSize, f"{instance['MemoryInfo']['SizeInMiB']}MiB"
+                    ),
+                )
+            )
+    return list_instances
+
+
+def closest_instance_policy(
+    ec2_instance: EC2Instance,
+    resources: Resources,
+) -> float:
+    if ec2_instance.cpus < resources.cpus or ec2_instance.ram < resources.ram:
+        return 0
+    # compute a score for all the instances that are above expectations
+    # best is the exact ec2 instance
+    cpu_ratio = float(ec2_instance.cpus - resources.cpus) / float(ec2_instance.cpus)
+    ram_ratio = float(ec2_instance.ram - resources.ram) / float(ec2_instance.ram)
+    return 100 * (1.0 - cpu_ratio) * (1.0 - ram_ratio)
+
+
+def find_best_fitting_ec2_instance(
+    available_ec2_instances: list[EC2Instance],
+    resources: Resources,
+    score_type=closest_instance_policy,
+) -> EC2Instance:
+    score_to_ec2_candidate: dict[float, EC2Instance] = OrderedDict(
+        sorted(
+            {
+                score_type(instance, resources): instance
+                for instance in available_ec2_instances
+            }.items(),
+            reverse=True,
+        )
     )
+    if not score_to_ec2_candidate:
+        raise Ec2InstanceNotFoundError(needed_resources=resources)
+    score, instance = next(iter(score_to_ec2_candidate.items()))
+    if score == 0:
+        raise Ec2InstanceNotFoundError(
+            needed_resources=resources, msg="no adequate EC2 instance found!"
+        )
+    return instance
 
 
-def create_ec2_client(settings: AwsSettings):
-    return boto3.client(
-        "ec2",
-        aws_access_key_id=settings.AWS_ACCESS_KEY_ID,
-        aws_secret_access_key=settings.AWS_SECRET_ACCESS_KEY,
-        region_name=settings.AWS_REGION_NAME,
+def _compose_user_data(docker_join_script: str) -> str:
+    return dedent(
+        f"""\
+#!/bin/bash
```

Comment:
To me it seems more clean and logical to put the shebang in the `docker_join_script` (thus add it to the str returned by `get_docker_swarm_join_script`). I guess your thinking is that you generate a **command** in `get_docker_swarm_join_script` and wrap it into a `.bash` **script** in a second step it `_compose_user_data`. If so, then at least the name of the function `_compose_user_data` I find unintuitive, maybe do `_generate_shell_script_from_command`. Or skip it alltogether and make `get_docker_swarm_join_script` return a properly validated bash script, with (pun intended) *the whole shebang* :D

Additionally, if this is possible, I would love to see a pytest that runs shellcheck on the final `.bash` file generated by `get_docker_swarm_join_script`, if you feel like it.

Reply:
since this is a AWS terminology and they themselves call it UserData (see AWS SDK)[https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/ec2.html#EC2.Client.run_instances](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/ec2.html#EC2.Client.run_instances)
I will keep it that way to avoid more confusion. The sheband is needed by this UserData only. and since I did not know whether we might need additional things run on start, for now I keep it so.

Concerning the shellcheck pytest script, I do not really see the time to code/gain ratio now since this is generated from docker now. I might change my mind in one of the next iterations though.

I will rename the other method to ```docker_join_bash_command```, I think that is clearer.